store/sqlstore/group_supplier.go

```diff
@@ -765,3 +765,92 @@ func groupSyncableToGroupChannel(groupSyncable *model.GroupSyncable) *groupChann
 		ChannelId:     groupSyncable.SyncableId,
 	}
 }
+
+// TeamMembersToRemove returns all team members that should be removed based on group constraints.
+func (s *SqlSupplier) TeamMembersToRemove(ctx context.Context, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	sql := `
+		SELECT
+			TeamMembers.*
+		FROM
+			TeamMembers
+			JOIN Teams ON Teams.Id = TeamMembers.TeamId
+		WHERE
+			TeamMembers.DeleteAt = 0
+			AND Teams.DeleteAt = 0
+			AND Teams.GroupConstrained = TRUE
+			AND (TeamMembers.TeamId, TeamMembers.UserId)
+			NOT IN (
+				SELECT
+					Teams.Id AS TeamId, GroupMembers.UserId
+				FROM
+					Teams
+					JOIN GroupTeams ON GroupTeams.TeamId = Teams.Id
+					JOIN UserGroups ON UserGroups.Id = GroupTeams.GroupId
+					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
+				WHERE
+					Teams.GroupConstrained = TRUE
+					AND GroupTeams.DeleteAt = 0
+					AND UserGroups.DeleteAt = 0
+					AND Teams.DeleteAt = 0
+					AND GroupMembers.DeleteAt = 0
+				GROUP BY
+					Teams.Id,
+					GroupMembers.UserId)`
+
+	var teamMembers []*model.TeamMember
+
+	_, err := s.GetReplica().Select(&teamMembers, sql)
+	if err != nil {
+		result.Err = model.NewAppError("SqlGroupStore.TeamMembersToRemove", "store.select_error", nil, "", http.StatusInternalServerError)
+	}
+
+	result.Data = teamMembers
+
+	return result
+}
+
+// ChannelMembersToRemove returns all channel members that should be removed based on group constraints.
+func (s *SqlSupplier) ChannelMembersToRemove(ctx context.Context, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	sql := `
+		SELECT
+			ChannelMembers.*
+		FROM
+			ChannelMembers
+			JOIN Channels ON Channels.Id = ChannelMembers.ChannelId
+		WHERE
+			Channels.DeleteAt = 0
+			AND Channels.GroupConstrained = TRUE
+			AND (ChannelMembers.ChannelId, ChannelMembers.UserId)
+			NOT IN (
+				SELECT
+					Channels.Id AS ChannelId, GroupMembers.UserId
+				FROM
+					Channels
+					JOIN GroupChannels ON GroupChannels.ChannelId = Channels.Id
+					JOIN UserGroups ON UserGroups.Id = GroupChannels.GroupId
+					JOIN GroupMembers ON GroupMembers.GroupId = UserGroups.Id
+				WHERE
+					Channels.GroupConstrained = TRUE
+					AND GroupChannels.DeleteAt = 0
+					AND UserGroups.DeleteAt = 0
+					AND Channels.DeleteAt = 0
+					AND GroupMembers.DeleteAt = 0
+				GROUP BY
+					Channels.Id,
+					GroupMembers.UserId)`
```

Comment:
See https://github.com/mattermost/mattermost-server/pull/10490/files#r270404176 for an explanation of this query. Everything applies to this one too.